src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -1143,16 +1143,22 @@ abstract class ClassfileParser {
         (newStub(name.toTypeName), newStub(name.toTermName))
       } else {
         val cls = owner.newClass(name.toTypeName, NoPosition, sflags)
-        if (cls.isProtected) {
-          cls.privateWithin = cls.enclosingPackage
-        }
         val mod = owner.newModule(name.toTermName, NoPosition, sflags)
         val completer = new loaders.ClassfileLoader(file, cls, mod)
         cls setInfo completer
         mod setInfo completer
         mod.moduleClass setInfo loaders.moduleClassLoader
         cls.associatedFile = file
         mod.moduleClass.associatedFile = file
+
+        /**
+          * need to set privateWithin here because the classfile of a nested protected class is public in bytecode,
+          * so propagatePackageBoundary will not set it when the symbols are completed
+          */
+        if (jflags.isProtected) {
+          cls.privateWithin = cls.enclosingPackage
```

Comment:
I think we also need to set `mod.privateWithin = ...`, see the test case in my comment https://github.com/scala/scala/pull/7323#pullrequestreview-163307986. Add an instance and a static member to the two nested java classes `StaticClass` and `ProtectedStaticClass`, and access these members from Scala code.